File: providers/python/autobyteus_voice_provider/normalization.py
from __future__ import annotations

import json
import re
import unicodedata
from pathlib import Path

SPACE = re.compile(r"\s+")
HAN_SPACE = re.compile(r"(?<=[\u3400-\u9fff])\s+(?=[\u3400-\u9fff])")
BEFORE_PUNCT = re.compile(r"\s+([,，.!！？?。])")
AFTER_ZH_PUNCT = re.compile(r"([，。！？])\s+")
END = object()


def _trie(dictionary: dict[str, str]) -> dict:
    root: dict = {}
    for source, replacement in dictionary.items():
        node = root
        for character in source:
            node = node.setdefault(character, {})
        node[END] = replacement
    return root


def _match(value: str, offset: int, trie: dict):
    node = trie
    best = None
    for index in range(offset, len(value)):
        node = node.get(value[index])
        if node is None:
            break
        if END in node:
            best = index + 1, node[END]
    return best


def _convert(value: str, trie: dict) -> str:
    output = []
    offset = 0
    while offset < len(value):
        matched = _match(value, offset, trie)
        if matched:
            offset, replacement = matched
            output.append(replacement)
        else:
            output.append(value[offset])
            offset += 1
    return "".join(output)


def _segments(value: str, trie: dict) -> list[str]:
    result = []
    unmatched = []
    offset = 0
    while offset < len(value):
        matched = _match(value, offset, trie)
        if matched:
            if unmatched:
                result.append("".join(unmatched))
                unmatched = []
            end, _ = matched
            result.append(value[offset:end])
            offset = end
        else:
            unmatched.append(value[offset])
            offset += 1
    if unmatched:
        result.append("".join(unmatched))
    return result


def _dictionary(value) -> dict[str, str]:
    if not isinstance(value, dict) or any(
        not isinstance(source, str)
        or not source
        or not isinstance(replacement, str)
        for source, replacement in value.items()
    ):
        raise ValueError("normalization-data-invalid")
    return value


class TranscriptNormalizer:
    def __init__(self, profile_id: str, mapping_path: Path | None = None):
        self.profile_id = profile_id
        self.normalizer = None
        self.segmenter = None
        self.stages = []
        if profile_id in ("chinese", "auto"):
            if mapping_path is None or not mapping_path.is_file():
                raise ValueError("normalization-data-missing")
            value = json.loads(mapping_path.read_text("utf-8"))
            expected = {"schemaVersion", "source", "normalization", "segmentation", "conversionStages"}
            source = value.get("source") if isinstance(value, dict) else None
            stages = value.get("conversionStages") if isinstance(value, dict) else None
            if (
                not isinstance(value, dict)
                or set(value) != expected
                or value["schemaVersion"] != 1
                or not isinstance(source, dict)
                or source.get("package") != "opencc-js"
                or source.get("version") != "1.4.1"
                or source.get("configuration") != "twp-to-cn"
                or not isinstance(stages, list)
                or len(stages) != 2
            ):
                raise ValueError("normalization-data-invalid")
            self.normalizer = _trie(_dictionary(value["normalization"]))
            self.segmenter = _trie(_dictionary(value["segmentation"]))
            self.stages = [_trie(_dictionary(stage)) for stage in stages]

    def _traditional_to_simplified(self, value: str) -> str:
        value = _convert(value, self.normalizer)
        output = []
        for segment in _segments(value, self.segmenter):
            for stage in self.stages:
                segment = _convert(segment, stage)
            output.append(segment)
        return "".join(output)

    def normalize(self, raw: str) -> str:
        value = SPACE.sub(" ", unicodedata.normalize("NFKC", raw)).strip()
        if self.profile_id in ("chinese", "auto") and re.search(r"[\u3400-\u9fff]", value):
            value = self._traditional_to_simplified(value)
            value = re.sub(r"(?<!\d),|,(?!\d)", "，", value)
            value = re.sub(r"(?<!\d)\.|\.(?!\d)", "。", value)
            value = value.replace("!", "！").replace("?", "？")
            value = HAN_SPACE.sub("", value)
            value = BEFORE_PUNCT.sub(r"\1", value)
            value = AFTER_ZH_PUNCT.sub(r"\1", value)
        return value

File: benchmark/scoring/normalization.mjs
import fs from "node:fs";
import path from "node:path";

const mapping = JSON.parse(
  fs.readFileSync(
    path.resolve(
      import.meta.dirname,
      "../../contracts/normalization/twp-to-cn-v1.json",
    ),
    "utf8",
  ),
);
assertMapping(mapping);
const normalizer = trie(mapping.normalization);
const segmenter = trie(mapping.segmentation);
const conversionStages = mapping.conversionStages.map(trie);

function assertMapping(value) {
  if (
    !value ||
    value.schemaVersion !== 1 ||
    Object.keys(value).sort().join(",") !==
      "conversionStages,normalization,schemaVersion,segmentation,source" ||
    value.source?.package !== "opencc-js" ||
    value.source?.version !== "1.4.1" ||
    value.source?.configuration !== "twp-to-cn" ||
    !Array.isArray(value.conversionStages) ||
    value.conversionStages.length !== 2
  )
    throw new Error("Invalid canonical normalization mapping.");
  for (const dictionary of [
    value.normalization,
    value.segmentation,
    ...value.conversionStages,
  ])
    if (
      !dictionary ||
      Array.isArray(dictionary) ||
      Object.entries(dictionary).some(
        ([source, replacement]) => !source || typeof replacement !== "string",
      )
    )
      throw new Error("Invalid canonical normalization dictionary.");
}

function trie(dictionary) {
  const root = new Map();
  for (const [source, replacement] of Object.entries(dictionary)) {
    let node = root;
    for (const character of source) {
      if (!node.has(character)) node.set(character, new Map());
      node = node.get(character);
    }
    node.replacement = replacement;
  }
  return root;
}

function match(input, offset, root) {
  let node = root;
  let best;
  for (let index = offset; index < input.length; index += 1) {
    node = node.get(input[index]);
    if (!node) break;
    if (Object.hasOwn(node, "replacement"))
      best = { end: index + 1, value: node.replacement };
  }
  return best;
}

function convert(value, dictionary) {
  const input = [...value];
  const output = [];
  for (let index = 0; index < input.length; ) {
    const matched = match(input, index, dictionary);
    if (matched) {
      output.push(matched.value);
      index = matched.end;
    } else {
      output.push(input[index]);
      index += 1;
    }
  }
  return output.join("");
}

function segments(value) {
  const input = [...value];
  const result = [];
  let unmatched = "";
  for (let index = 0; index < input.length; ) {
    const matched = match(input, index, segmenter);
    if (matched) {
      if (unmatched) result.push(unmatched);
      unmatched = "";
      result.push(input.slice(index, matched.end).join(""));
      index = matched.end;
    } else {
      unmatched += input[index];
      index += 1;
    }
  }
  if (unmatched) result.push(unmatched);
  return result;
}

function traditionalToSimplified(value) {
  const normalized = convert(value, normalizer);
  return segments(normalized)
    .map((segment) =>
      conversionStages.reduce(
        (converted, dictionary) => convert(converted, dictionary),
        segment,
      ),
    )
    .join("");
}

export function normalizeTranscript(raw, profileId) {
  let value = raw.normalize("NFKC").replace(/\s+/gu, " ").trim();
  if (profileId === "english") return value;
  if (profileId !== "chinese" && profileId !== "auto")
    throw new Error("Unknown normalization profile.");
  if (!/[\p{Script=Han}]/u.test(value)) return value;
  value = traditionalToSimplified(value)
    .replace(/(?<!\d),|,(?!\d)/g, "，")
    .replace(/(?<!\d)\.|\.(?!\d)/g, "。")
    .replace(/!/g, "！")
    .replace(/\?/g, "？")
    .replace(/(?<=\p{Script=Han})\s+(?=\p{Script=Han})/gu, "")
    .replace(/\s+([，。！？])/gu, "$1")
    .replace(/([，。！？])\s+/gu, "$1");
  return value;
}

export function englishWerUnits(value) {
  return normalizeTranscript(value, "english")
    .toLowerCase()
    .replace(/[^\p{L}\p{N}' ]/gu, " ")
    .split(/\s+/)
    .filter(Boolean);
}
